### PR TITLE
test: add client tests

### PR DESCRIPTION
**Pull request by juliusmarminge**
more stuff to increase confident in Effect PR. we're going places

CleanShot 2024-03-06 at 20 18 35

### CODE REVIEW

File: patches/msw+2.2.2.patch
Comment (juliusmarminge) [inline patches/msw+2.2.2.patch:1]: msw doesn't really like that we don't use global fetch but pass it as args... so this makes it so we only intercept XHR and mock the other requests by passing the `mockExternalRequests` function as the fetch option...

File: packages/uploadthing/test/sdk.test.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ describe("uploadFiles", () => {
     );
     expect(fetchMock).toHaveBeenNthCalledWith(
       2,
-      "https://bucket.s3.amazonaws.com/abc-123.txt",
+      "https://bucket.s3.amazonaws.com",
```
Comment (juliusmarminge) [inline packages/uploadthing/test/sdk.test.ts:55]: post urls sends the key in body

File: packages/uploadthing/test/__test-helpers.ts [hunk 2]
```diff
@@ -1,17 +1,30 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (markflorkowski) [inline packages/uploadthing/test/__test-helpers.ts:1]: only needed for line 129?

Comment (juliusmarminge) [reply packages/uploadthing/test/__test-helpers.ts]: fixed

File: packages/uploadthing/test/client.test.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        url: "https://utfs.io/f/abc-123.txt",
+      },
+    ]);
+
+    expect(s3Mock).toHaveBeenCalledOnce();
+    expect(s3Mock).toHaveBeenCalledWith({}, expect.any(Request));
+
+    expect(middlewareMock).toHaveBeenCalledOnce();
+    expect(onErrorMock).not.toHaveBeenCalled();
+    // expect(uploadCompleteMock).toHaveBeenCalledOnce();
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply packages/uploadthing/test/client.test.ts:90]: Is there some good way to store context from the test? So we can simulate the UTApi even better and store these when uploading, and have pollupload read from that? Have to be isolated to a single test

[the remark this answers is not in the document]
Comment (juliusmarminge) [reply packages/uploadthing/test/client.test.ts]: took a while, but got it working

**Pull request merged**